src/Tasks/dependency-check-build-task/dependency-check-build-task.ts

```diff
@@ -207,9 +200,15 @@ async function run() {
                 let fileExt = path.extname(filePath);
                 logDebug(`Attachment name: ${fileName}`);
                 logDebug(`Attachment path: ${filePath}`);
-                logDebug(`Attachment type: ${fileExt}`); 
-                console.log(`##vso[task.addattachment type=dependencycheck-artifact;name=${fileName};]${filePath}`);
-                console.log(`##vso[artifact.upload containerfolder=dependency-check;artifactname=Dependency Check;]${filePath}`);
+                logDebug(`Attachment type: ${fileExt}`);
+
+                tl.uploadArtifact('dependency-check', filePath, 'Dependency Check')
+
+                // To display the SARIF report in Azure DevOps with the SARIF SAST Scans Tab extension, the report must me in the CodeAnalysisLogs artifact 
+                if (fileExt.toLowerCase() === '.sarif') {
+                    logDebug(`Uploaded SARIF attachment: ${fileName}`);
+                    tl.uploadArtifact('OWASPDependencyCheck', `${filePath}`, 'CodeAnalysisLogs')
```

Comment:
If for whatever reason you need to re-run the job that contains this task and this task succeeded the first time, then this and line 205 will  fail as the artifacts have already been uploaded and they are immutable--which is pretty annoying. In my custom tasks I usually offer either a way to customize the artifact name in full *or* an "artifact suffix" that can be passed to the task and appended to the artifact names (which I then usually set to `_System.JobAttempt`) 

Similarly, because there's issues with uploading to an existing artifact (unless done all at once), I think it might be preferable (or at least customizable) to *not* upload to the CodeAnalysisLogs artifact directly from this task. Otherwise, if there were sarif reports generated by other tasks/tools then this task running first would preclude a generalized `publish artifact` task from being able to upload those files (or vice versa)

Just some thoughts...

Reply:
Hi, @pinkfloydx33, I understand what you are saying but can you explain how you can rerun a non-deployment job? I prefer using guide for unique file names in this case.

Reply:
You can rerun any job in an stage, failed or not, build, deploy or otherwise. It's more obvious if a task fails where you get a "re-run failed jobs" button, but you can technically rerun an entire stage once it's finished (though less likely you'd do this). It's all definitely there with Yaml pipelines, I'm not sure about classic. 

Where the problem really arises is during an intermittent failure. Say you have a task *after* this one that failed due to network issue. So you click "rerun failed jobs". Only this time it will fail on **this** task because you're trying to upload to an artifact that already exists, which is disallowed. Now the only way forward is to create an entirely new run of the pipeline.

Reply:
Agree... but to be honest, I don't think this is in scope of this PR. The extension needs some rework where the most important thing is using the tool cache. I am willing to help but let's see what the maintainers think about this.

Making the file upload independent is also important; that starts with unique filenames or artifactnames with the correct name.

Reply:
Valid discussion here. 

* I agree that we should provide an input parameter that enables / disables the auto-uploading of artifacts. 
* I also agree that we should make the artifacts support multiple attempts if possible. Perhaps something like "${System.JobAttempt}_dependencycheck.${extension}" would be sufficient to fix that problem.

Reply:
I will concede that the naming of the artifact is likely outside of the scope of the PR.

Automatically uploading the sarif artifact, however needs to (IMO) be optional. Otherwise you have introduced a bug that will not surface within the task itself, but rather in subsequent tasks when the user attempts to upload other reports to the `CodeAnalysisLogs` artifact--which may be difficult to diagnose. Users already using the viewer for other reports will need to decide whether they want to use it for the results of this task **only**, or forego generating sarif files via this task altogether so they can continue to view the reports they already had. There's got to be a middle ground.

Note: this wouldn't impact me. But I'm sure there's someone out there currently using the viewer that would like to add this in, only to be surprised they couldn't. I'm merely replying as a custom task author (and the person who reported the node10 runner issue).

Reply:
> * I also agree that we should make the artifacts support multiple attempts if possible. Perhaps something like "${System.JobAttempt}_dependencycheck.${extension}" would be sufficient to fix that problem.

The way I typically do it in my own tasks (if not letting the user pass the full name and thus control a potential suffix/prefix themselves) is to use the System.JobAttempt. However I usually add a bit of logic to see if it's Attempt=1 and *not* incorporate it, since it's typically noise and the first attempt should (hopefully) succeed the majority of the time. But either way works, just offering some notes from experience.

Reply:
The real worry is not the job attempts. The real worry is multi job / multi stage pipelines. The files are just overwritten when the extension runs twice in the same pipeline. Again: nothing to do with my PR but I will dive into this.

Reply:
boolean options for upload reports / upload SARIF reports are processed

Reply:
Okay.. this is.... pain.

We have: multi job, multi stage, job attempts and stage attempts. So 4 variables that influence the name of a unique report name. Next there is strange behaviour of Azure DevOps, For multistage pipelines, when a stage has a name the stage name can be found with System.StageName. The job name however, stays '__default', you can have two jobs with the name '__default' within the same stage, again not creating an unique report name

Appending things like attempts, stage name and job names is not explainable to users since the buildup of the appendix cannot follow an easy explainable structure.

Reply:
Okay.. this is.... pain.

We have: multi job, multi stage, job attempts and stage attempts. So 4 variables that can trigger uploading a report with the same name, overriding the old.

Next there is strange behaviour of Azure DevOps, For multistage pipelines, when a stage has a name the stage name can be found with System.StageName. The job name however, stays '__default', you can have two jobs with the name '__default' within the same stage, again not creating an unique report name.  This also applies for two jobs in a classic pipeline.  Using a displayname is not unique.

Appending things like attempts, stage names and job names is not explainable to users since the buildup of the appendix cannot follow an easy explainable structure. Why include ‘__default’? 

To be honest, I don’t see a way out of this besides putting the JobId (guid) in the filename to generate an unique report name. And if we do this, the behaviour changes of this task. For the Gitleaks extension this was my approach to generate an unique report

Simplest way to reproduce: Two jobs in a classic build pipeline. 

@pinkfloydx33 @ejohn20 what do you both think?

Reply:
I think we've reached a point where it's probably worth opening a a separate issue. These edges cases have not come up yet, and perhaps it's not really that big of a deal. 

That said, I think the `jobid` would be fine to include in the file name if it's unique.

Reply:
#94 opened. I mark this as resolved for this PR